### ISSUE TITLE
Recursive processing with many PDF files fail with "Too many open files"

### ISSUE THREAD

**Issue by ravn** [veraPDF/veraPDF-apps#133]
When running

 ~/verapdf/verapdf -f 1b -r --format folder

(where folder contain 4921 PDF-files), VeraPDF fails with the following exception

```
    org.verapdf.core.VeraPDFException: Couldn't find file: dl1_20161201_rt1/nvk/pages/20161201_nordvestnytkalundborg_section03_page028_nvk20161201x13#0028.pdf to process.
	at org.verapdf.processor.ProcessorImpl.process(ProcessorImpl.java:100)
	at org.verapdf.processor.BatchFileProcessor.processItem(BatchFileProcessor.java:75)
	at org.verapdf.processor.BatchFileProcessor.processDir(BatchFileProcessor.java:63)
	at org.verapdf.processor.BatchFileProcessor.processDir(BatchFileProcessor.java:61)
	at org.verapdf.processor.BatchFileProcessor.processDir(BatchFileProcessor.java:61)
	at org.verapdf.processor.BatchFileProcessor.processContainer(BatchFileProcessor.java:38)
	at org.verapdf.processor.AbstractBatchProcessor.process(AbstractBatchProcessor.java:73)
	at org.verapdf.cli.VeraPdfCliProcessor.processBatch(VeraPdfCliProcessor.java:158)
	at org.verapdf.cli.VeraPdfCliProcessor.processDir(VeraPdfCliProcessor.java:121)
	at org.verapdf.cli.VeraPdfCliProcessor.processPaths(VeraPdfCliProcessor.java:100)
	at org.verapdf.cli.VeraPdfCli.main(VeraPdfCli.java:79)
Caused by: java.io.FileNotFoundException: dl1_20161201_rt1/nvk/pages/20161201_nordvestnytkalundborg_section03_page028_nvk20161201x13#0028.pdf (Too many open files)
	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(FileInputStream.java:195)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at org.verapdf.processor.ProcessorImpl.process(ProcessorImpl.java:97)
	... 10 more

```

The output indicates that 986 files had been processed.  Environment Ubuntu 16.10 with Oracle Java 1.8.0_102 64bit.

**Comment by ravn** [comment]
This is the GL edition of 0.28

**Comment by BezrukovM** [comment]
@ravn could you please provide us exact version of veraPDF app which you are using (GreenField or PDFBox based and exact version with third number) and full command for running (--format requires one of the following arguments: mrr, xml or text).

**Comment by ravn** [comment]
My apologies.  My editing the command line apparently resulted in me removing the format argument also.


In the http://verapdf.org/2016/12/21/verapdf-0-28-released/ announcement there is a link to http://downloads.verapdf.org/gf/verapdf-gf-installer.zip which appears to be 0.28.4-GF  in the current download.  This was installed with just clicking Next all the time resulting in ~/verapdf being created and populated.

Repeating the exercise gives the following:

```

~/verapdf/verapdf -f 1b -r --format mrr dl1_20161201_rt1/ >/tmp/1.txt


org.verapdf.core.VeraPDFException: Couldn't find file: dl1_20161201_rt1/drs/pages/20161201_ringkjoebingamtsdagblad_section04_page007_drs20161201x14#0007.pdf to process.
	at org.verapdf.processor.ProcessorImpl.process(ProcessorImpl.java:100)
	at org.verapdf.processor.BatchFileProcessor.processItem(BatchFileProcessor.java:75)
	at org.verapdf.processor.BatchFileProcessor.processDir(BatchFileProcessor.java:63)
	at org.verapdf.processor.BatchFileProcessor.processDir(BatchFileProcessor.java:61)
	at org.verapdf.processor.BatchFileProcessor.processDir(BatchFileProcessor.java:61)
	at org.verapdf.processor.BatchFileProcessor.processContainer(BatchFileProcessor.java:38)
	at org.verapdf.processor.AbstractBatchProcessor.process(AbstractBatchProcessor.java:73)
	at org.verapdf.cli.VeraPdfCliProcessor.processBatch(VeraPdfCliProcessor.java:158)
	at org.verapdf.cli.VeraPdfCliProcessor.processDir(VeraPdfCliProcessor.java:121)
	at org.verapdf.cli.VeraPdfCliProcessor.processPaths(VeraPdfCliProcessor.java:100)
	at org.verapdf.cli.VeraPdfCli.main(VeraPdfCli.java:79)
Caused by: java.io.FileNotFoundException: dl1_20161201_rt1/drs/pages/20161201_ringkjoebingamtsdagblad_section04_page007_drs20161201x14#0007.pdf (Too many open files)
	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(FileInputStream.java:195)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at org.verapdf.processor.ProcessorImpl.process(ProcessorImpl.java:97)
	... 10 more

```


Also while preparing this comment (where I looked at stdout too), I noticed several stack traces being printed on the lines of:

```
WARNING: Exception caught when validaing item
org.verapdf.core.ValidationException: Caught unexpected runtime exception during validation
	at org.verapdf.pdfa.validation.validators.BaseValidator.validate(BaseValidator.java:89)
	at org.verapdf.processor.ProcessorImpl.validate(ProcessorImpl.java:176)
	at org.verapdf.processor.ProcessorImpl.process(ProcessorImpl.java:119)
	at org.verapdf.processor.ProcessorImpl.process(ProcessorImpl.java:98)
	at org.verapdf.processor.BatchFileProcessor.processItem(BatchFileProcessor.java:75)
	at org.verapdf.processor.BatchFileProcessor.processDir(BatchFileProcessor.java:63)
	at org.verapdf.processor.BatchFileProcessor.processDir(BatchFileProcessor.java:61)
	at org.verapdf.processor.BatchFileProcessor.processDir(BatchFileProcessor.java:61)
	at org.verapdf.processor.BatchFileProcessor.processContainer(BatchFileProcessor.java:38)
	at org.verapdf.processor.AbstractBatchProcessor.process(AbstractBatchProcessor.java:73)
	at org.verapdf.cli.VeraPdfCliProcessor.processBatch(VeraPdfCliProcessor.java:158)
	at org.verapdf.cli.VeraPdfCliProcessor.processDir(VeraPdfCliProcessor.java:121)
	at org.verapdf.cli.VeraPdfCliProcessor.processPaths(VeraPdfCliProcessor.java:100)
	at org.verapdf.cli.VeraPdfCli.main(VeraPdfCli.java:79)
Caused by: java.lang.NullPointerException
	at org.verapdf.pd.font.truetype.TrueTypeFontParser.getCmapTable(TrueTypeFontParser.java:129)
	at org.verapdf.pd.font.truetype.TrueTypeFontProgram.containsCode(TrueTypeFontProgram.java:64)
	at org.verapdf.gf.model.impl.operator.textshow.GFOpTextShow.getUsedGlyphs(GFOpTextShow.java:143)
	at org.verapdf.gf.model.impl.operator.textshow.GFOpTextShow.getLinkedObjects(GFOpTextShow.java:96)
	at org.verapdf.gf.model.impl.operator.textshow.GFOp_TJ_Big.getLinkedObjects(GFOp_TJ_Big.java:39)
	at org.verapdf.pdfa.validation.validators.BaseValidator.addAllLinkedObjects(BaseValidator.java:220)
	at org.verapdf.pdfa.validation.validators.BaseValidator.checkNext(BaseValidator.java:165)
	at org.verapdf.pdfa.validation.validators.BaseValidator.validate(BaseValidator.java:116)
	at org.verapdf.pdfa.validation.validators.BaseValidator.validate(BaseValidator.java:87)
	... 13 more

```

**Comment by ravn** [comment]
We have been busy with other work.  I have now returned to this issue as we would like to put our VeraPDF component in production and it appears not to be fixed in the 1.4.something artifact we use in code, or in the 1.6.3 standalone version I have just downloaded.     I cannot reopen this issue, so I'll open a new one.
